### PR TITLE
[WIP] Fix(Core/Script) Temple of Ahn'Qiraj

### PR DESCRIPTION
**Pull request by imxari**
# Work in progress update for Temple of Ahn'Qiraj
Any feedback is welcomed, if you find an issue testing any of the converted bosses listed below please do let me know.

## CHANGES PROPOSED:
-  All bosses will be converted over from uint32 timers to use the newer EventMap system.
-  Boss code will be optimised and shortened where possible.

## ISSUES ADDRESSED:
- None

## TESTS PERFORMED:
- Building the core
- Testing each boss individually in-game after each change.

## HOW TO TEST THE CHANGES:
- Build the core using my PR
- Teleport to the Temple of Ahn'Qiraj
- Test bosses that are marked as converted.

## Converted Bossess
- [x] Bug Trio
- [x] Fankriss
- [x] Huhuran
- [x] Sartura
- [x] Skeram
- [x] Twin Emperors
- [x] Viscidus
- [ ] C'Thun
- [x] Ouro: Has already been addressed by another PR. https://github.com/azerothcore/azerothcore-wotlk/pull/2997

## Target branch(es):
- [x] Master

 
## How to test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here in the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

### CODE REVIEW

**Comment by Yehonal** [comment]
Cool! I love Ahn Quiraj!

It will be amazing to play a fully working AQ experience with the timewalking module I'm going to release soon :D

Keep up the good work!

**Comment by h0lysp4nk** [comment]
Almost done revamping the Temple of Ahn'Qiraj, just have C'Thun left. 😄

File: src/server/scripts/Kalimdor/TempleOfAhnQiraj/boss_bug_trio.cpp [hunk 1]
```diff
@@ -47,6 +47,28 @@ enum Events
...
+    {
+        if (Creature* vem = ObjectAccessor::GetCreature(*me, instance->GetData64(DATA_VEM)))
+            if (vem->GetGUID() != me->GetGUID())
+                vem->GetAI()->AttackStart(pUnit);
+        if (Creature* kri = ObjectAccessor::GetCreature(*me, instance->GetData64(DATA_KRI)))
+            if (kri->GetGUID() != me->GetGUID())
+                kri->GetAI()->AttackStart(pUnit);
+        if (Creature* yauj = ObjectAccessor::GetCreature(*me, instance->GetData64(DATA_YAUJ)))
+            if (yauj->GetGUID() != me->GetGUID())
+                yauj->GetAI()->AttackStart(pUnit);
```
Comment (Helias) [inline src/server/scripts/Kalimdor/TempleOfAhnQiraj/boss_bug_trio.cpp:65]: little suggestion: you could merge the conditions into **one if** here:
```C++
        if (Creature* vem = ObjectAccessor::GetCreature(*me, instance->GetData64(DATA_VEM)) && vem->GetGUID() != me->GetGUID())
            vem->GetAI()->AttackStart(pUnit);
        if (Creature* kri = ObjectAccessor::GetCreature(*me, instance->GetData64(DATA_KRI)) && kri->GetGUID() != me->GetGUID())
            kri->GetAI()->AttackStart(pUnit);
        if (Creature* yauj = ObjectAccessor::GetCreature(*me, instance->GetData64(DATA_YAUJ)) && yauj->GetGUID() != me->GetGUID())
            yauj->GetAI()->AttackStart(pUnit);
```

File: src/server/scripts/Kalimdor/TempleOfAhnQiraj/boss_fankriss.cpp [hunk 2]
```diff
@@ -41,155 +42,115 @@ class boss_fankriss : public CreatureScript
...
+                        {
+                            summoned.Summon(pCreature);
+                            pCreature->AI()->AttackStart(pUnit);
+                        }
+                        if (Creature* pCreature = me->SummonCreature(15962, pUnit->GetPositionX() - 5, pUnit->GetPositionY() + 5, pUnit->GetPositionZ(), 0, TEMPSUMMON_TIMED_DESPAWN_OUT_OF_COMBAT, 15000))
+                        {
+                            summoned.Summon(pCreature);
+                            pCreature->AI()->AttackStart(pUnit);
                         }
+                        break;
```
Comment (Helias) [inline src/server/scripts/Kalimdor/TempleOfAhnQiraj/boss_fankriss.cpp:141]: The `pCreature` will be only one? or there is a case where you will `summon` 4 creatures?
if not, you can write this more short  like this way I think:
```C++
Creature* pCreature;

if (
  pCreature = me->SummonCreature(15962, pUnit->GetPositionX() - 3, pUnit->GetPositionY() - 3, pUnit->GetPositionZ(), 0, TEMPSUMMON_TIMED_DESPAWN_OUT_OF_COMBAT, 15000) || 
  pCreature = me->SummonCreature(15962, pUnit->GetPositionX() - 3, pUnit->GetPositionY() + 3, pUnit->GetPositionZ(), 0, TEMPSUMMON_TIMED_DESPAWN_OUT_OF_COMBAT, 15000) ||
  pCreature = me->SummonCreature(15962, pUnit->GetPositionX() - 5, pUnit->GetPositionY() - 5, pUnit->GetPositionZ(), 0, TEMPSUMMON_TIMED_DESPAWN_OUT_OF_COMBAT, 15000) ||
  pCreature = me->SummonCreature(15962, pUnit->GetPositionX() - 5, pUnit->GetPositionY() + 5, pUnit->GetPositionZ(), 0, TEMPSUMMON_TIMED_DESPAWN_OUT_OF_COMBAT, 15000)
) {
    summoned.Summon(pCreature);
    pCreature->AI()->AttackStart(pUnit);
}
```

but I am not sure if one of these if exclude the others, if they were several `if / else if` I would be sure about this suggestion

File: src/server/scripts/Kalimdor/TempleOfAhnQiraj/boss_twinemperors.cpp [hunk 3]
```diff
@@ -48,531 +48,418 @@ enum Misc
...
-        }
+
+        if (Creature* pBoss = GetOtherBoss())
+            if (pBoss->IsInCombat() == false)
+                if (ScriptedAI* pBossAI = CAST_AI(ScriptedAI, pBoss->AI()))
+                {
+                    DoPlaySoundToSet(me, IAmVeklor() ? SOUND_VL_AGGRO : SOUND_VN_AGGRO);
+                    pBossAI->AttackStart(pUnit);
+                    pBossAI->DoZoneInCombat();
+                }
```
Comment (Helias) [inline src/server/scripts/Kalimdor/TempleOfAhnQiraj/boss_twinemperors.cpp:137]: Here you could merge the if into one if with all conditions:
```C++
        if (Creature* pBoss = GetOtherBoss() &&
            pBoss->IsInCombat() == false &&
            ScriptedAI* pBossAI = CAST_AI(ScriptedAI, pBoss->AI())
       ) {
                  .....
       }
```

File: src/server/scripts/Kalimdor/TempleOfAhnQiraj/boss_twinemperors.cpp [hunk 4]
```diff
@@ -48,531 +48,418 @@ enum Misc
...
-                if (Teleport_Timer > diff)
-                    Teleport_Timer -= diff;
-                else
-                    Teleport_Timer = 0;
-                return false;
+
+                if (pCreature->IsWithinDistInMap(me, ABUSE_BUG_RANGE))
+                    if (!pCreatureNearby || (rand() % 4) == 0)
+                        pCreatureNearby = pCreature;
             }
```
Comment (Helias) [inline src/server/scripts/Kalimdor/TempleOfAhnQiraj/boss_twinemperors.cpp:239]: You could merge this if conditions into one

File: src/server/scripts/Kalimdor/TempleOfAhnQiraj/boss_twinemperors.cpp [hunk 5]
```diff
@@ -48,531 +48,418 @@ enum Misc
...
-            if (me->IsWithinDistInMap(who, PULL_RANGE) && me->GetDistanceZ(who) <= /*CREATURE_Z_ATTACK_RANGE*/7 /*there are stairs*/)
-            {
-                //if (who->HasStealthAura())
-                //    who->RemoveSpellsCausingAura(SPELL_AURA_MOD_STEALTH);
-                AttackStart(who);
-            }
-        }
+        if (me->_CanDetectFeignDeathOf(pUnit) && me->CanCreatureAttack(pUnit))
+            if (me->IsWithinDistInMap(pUnit, PULL_RANGE) && me->GetDistanceZ(pUnit) <= 7)
+                AttackStart(pUnit);
```
Comment (Helias) [inline src/server/scripts/Kalimdor/TempleOfAhnQiraj/boss_twinemperors.cpp:250]: same here

File: src/server/scripts/Kalimdor/TempleOfAhnQiraj/boss_twinemperors.cpp [hunk 6]
```diff
@@ -48,531 +48,418 @@ enum Misc
...
         }
 
         void UpdateAI(uint32 diff)
         {
-            //Return since we have no target
-            if (!UpdateVictim())
-                return;
-
-            if (!TryActivateAfterTTelep(diff))
+            if (UpdateVictim() == false)
```
Comment (Helias) [inline src/server/scripts/Kalimdor/TempleOfAhnQiraj/boss_twinemperors.cpp:291]: you could write this like `if (!UpdateVictim())`

File: src/server/scripts/Kalimdor/TempleOfAhnQiraj/boss_twinemperors.cpp [hunk 7]
```diff
@@ -48,531 +48,418 @@ enum Misc
...
+                    if (Creature* pCreature = RespawnNearbyBugsAndGetOne())
+                        MutateBug(pCreature);
+                    events.RepeatEvent(2000);
+                    break;
+                case EVENT_TELEPORT:
+                    TeleportToBrotherIfCan();
+                    events.ScheduleEvent(EVENT_TELEPORT_AFTER, 2000);
+                    events.RepeatEvent(30000);
+                case EVENT_TELEPORT_AFTER:
+                    if (AfterTeleport == true)
```
Comment (Helias) [inline src/server/scripts/Kalimdor/TempleOfAhnQiraj/boss_twinemperors.cpp:323]: probably you do not need to put `== true`, but you can convert the variable into boolean using this:
`if (!!AfterTeleport)`

File: src/server/scripts/Kalimdor/TempleOfAhnQiraj/boss_twinemperors.cpp [hunk 8]
```diff
@@ -48,531 +48,418 @@ enum Misc
...
-            if (!who)
+            if (!pUnit)
                 return;
 
-            if (who->isTargetableForAttack())
-            {
-                // VL doesn't melee
-                if (me->Attack(who, false))
+            if (pUnit->isTargetableForAttack())
+                if (me->Attack(pUnit, false))
```
Comment (Helias) [inline src/server/scripts/Kalimdor/TempleOfAhnQiraj/boss_twinemperors.cpp:455]: same here

**Comment by Helias** [comment]
I leave little comments in the code, they are only some suggestions but not mandatory for this PR.

Anyway good job @h0lysp4nk

File: src/server/scripts/Kalimdor/TempleOfAhnQiraj/boss_bug_trio.cpp (hunk 1 above)
Comment (h0lysp4nk) [reply src/server/scripts/Kalimdor/TempleOfAhnQiraj/boss_bug_trio.cpp:65]: That's true, will do. Thanks for the suggestion!, greatly appreciated :D

Comment (h0lysp4nk) [reply src/server/scripts/Kalimdor/TempleOfAhnQiraj/boss_bug_trio.cpp:65]: Aha, I remember why I did it the way it currently is. I tried something similar to the above when originally writing it but it gave a parsing error.

**Comment by cts17** [comment]
How goes C'thun? Should this be tested on the other bosses?

**Comment by imxari** [comment]
> How goes C'thun? Should this be tested on the other bosses?

I haven't had much time recently due to personal reasons, I'm working on it when I can but it's going quite slowly. I should hopefully have it done within the next month.

**Comment by imxari** [comment]
Heya,

Sorry I've been gone for so long, I've started working on this again. Going to fix Ouro and continue with C'Thun :)

**Comment by Yehonal** [comment]
Awesome! Keep up the good job!

Il lun 28 dic 2020, 13:59 imxari <notifications@github.com> ha scritto:

**Comment by FrancescoBorzi** [comment]
@imxari thanks for letting us know! keep it up

**Comment by Zoidwaffle** [comment]
I just wanted to say that I very much appreciate the work on fixing "classic" content, thanks!

**Comment by valens14** [comment]
I've been doing some scripting of AQ40 on my server as well.  Would you be interested in fixes to mechanics that aren't working?  Some things I have been able to fix like Bug Trio not respawning if you kill 1 or 2 of them and then wipe for example.  Others like C'thun's Dark Glare hitting a large cone area instead of a line I have not been able to fix.

**Comment by Zoidwaffle** [comment]
@valens14 - I can only speak for myself, but I prefer any fix that makes it closer to the "real" thing or at least something kinda working, over stuff that's not working at all... It does sound like this PR should already address all bosses except 2, but I haven't tested any of it (yet).

**Comment by Si1ker** [comment]
@imxari Hello! if you don't mind, I just wanted to know if this PR is still on progress

**Comment by Hacki95** [comment]
Hello @imxari . Are by chance in the discord server and could contact me on there (Hacki)? We are currently working on fixing every classic raid and I would like to ask you if you want to finish this PR/if you would be okay with someone else taking over your progress and finishing it.

**Comment by Kitzunu** [comment]
The completed parts should be cherry-picked to separate PRs, one per boss as the author seems to be inactive.

**Comment by temperrr** [comment]
Marked as abandoned since the author hasn't worked on it for a while.

**Pull request closed without merging**